app/src/main/java/org/mozilla/rocket/content/BottomBarItemAdapter.kt

```diff
@@ -0,0 +1,137 @@
+        TYPE_TAB_COUNTER -> TabCounterItem(type)
+        TYPE_MENU -> MenuItem(type)
+        TYPE_NEW_TAB -> ImageItem(type, R.drawable.action_add)
+        TYPE_SEARCH -> ImageItem(type, R.drawable.action_search)
+        TYPE_CAPTURE -> ImageItem(type, R.drawable.action_capture)
+        else -> error("Unexpected BottomBarItem ItemType: $type")
+    }
+
+    fun setNightMode(isNight: Boolean) {
+        items?.forEach {
+            val view = it.view
+            val type = it.type
+            when {
+                view is ThemedImageButton -> view.setNightMode(isNight)
+                type == TYPE_TAB_COUNTER -> (view as TabCounter).setNightMode(isNight)
+                type == TYPE_MENU -> view?.findViewById<ThemedImageButton>(R.id.btn_menu)?.setNightMode(isNight)
+            }
+        }
+    }
+
+    @JvmOverloads
+    fun setTabCount(count: Int, animationEnabled: Boolean = false) {
+        getItems(TYPE_TAB_COUNTER)
+                .map { (it as TabCounterItem).view as TabCounter }
+                .forEach {
+                    if (animationEnabled) {
+                        it.setCount(count)
+                    } else {
+                        it.setCountWithAnimation(count)
+                    }
+                }
+    }
+
+    fun setDownloadState(state: Int) {
+        getItems(TYPE_MENU).forEach {
+            val view = requireNotNull(it.view)
+            val stateIcon = view.findViewById<ImageView>(R.id.download_unread_indicator)
+            val downloadingAnimationView = view.findViewById<LottieAnimationView>(R.id.downloading_indicator)
+            when (state) {
+                DOWNLOAD_STATE_DEFAULT -> {
+                    stateIcon.visibility = View.GONE
+                    downloadingAnimationView.visibility = View.GONE
+                }
+                DOWNLOAD_STATE_DOWNLOADING -> {
+                    stateIcon.visibility = View.GONE
+                    downloadingAnimationView.apply {
+                        visibility = View.VISIBLE
+                        if (!downloadingAnimationView.isAnimating) {
+                            playAnimation()
+                        }
+                    }
+                }
+                DOWNLOAD_STATE_UNREAD -> {
+                    stateIcon.apply {
+                        visibility = View.VISIBLE
+                        setImageResource(R.drawable.notify_download)
+                    }
+                    downloadingAnimationView.visibility = View.GONE
+                }
+                DOWNLOAD_STATE_WARNING -> {
+                    stateIcon.apply {
+                        visibility = View.VISIBLE
+                        setImageResource(R.drawable.notify_notice)
+                    }
+                    downloadingAnimationView.visibility = View.GONE
+                }
+            }
+        }
+    }
+
+    private fun getItems(type: Int): List<BottomBarItem> =
```

Comment:
It's a little weird that the usage of getItems() inside the BottomBarItemAdapter class seems to allow multiple items with the same type, while the exposed findItem() can only return the first matched item. If you allow the users to put several items with the same type, you should also allow them to get all of them instead of only the first one.

Reply:
I have made BottomBarItemAdapter only accept non duplicated items now